Extend test-groups API to be able to fetch test group information from a build request.

dewei-zhu opened the pull request:
#### c87270cd9d151bf349a41486c919baa36786b8d5
<pre>
Extend test-groups API to be able to fetch test group information from a build request.
<a href="https://bugs.webkit.org/show_bug.cgi?id=270467">https://bugs.webkit.org/show_bug.cgi?id=270467</a>
<a href="https://rdar.apple.com/124024370">rdar://124024370</a>

Reviewed by Ryosuke Niwa.

Add `/api/test-groups?buildRequest=&lt;id&gt;` API so that we can query test-group information from a build request.

* Websites/perf.webkit.org/public/api/test-groups.php: Added support for `/api/test-groups?buildRequest=&lt;id&gt;`.
* Websites/perf.webkit.org/server-tests/api-test-groups.js: Added unit tests for new API.

Canonical link: <a href="https://commits.webkit.org/275648@main">https://commits.webkit.org/275648@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/698c8fe9f296d874e71a824c18d6fd1ad2a7979d

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/42424 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/21442 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/44818 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/45024 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/38541 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/24670 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/18790 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/35132 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/42998 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/18366 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/36521 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/16076 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/16016 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/46492 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/38609 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/37895 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/41807 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/17240 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/50/builds/14195 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/40415 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/18859 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/9486 "Built successfully and passed tests") | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/18921 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/18504 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rniwa commented on Websites/perf.webkit.org/public/api/test-groups.php:
```diff
@@ -15,13 +15,23 @@ function main($path) {
+        if (!$task_id && !$request_id)
+            exit_with_error('TaskOrBuildRequestIdNotSpecified');
+        if ($task_id && $request_id)
+            exit_with_error('CannotSpecifyBothTaskAndBuildRequest');
```
Maybe CannotSpecifyBothTaskAndBuildRequestIds?

rniwa commented on Websites/perf.webkit.org/server-tests/api-test-groups.js:
```diff
@@ -245,4 +246,55 @@ describe('/api/test-groups', function () {
+        });
+
+        it('should reject with "FailedToFetchTestGroups" if build request id is invalid', async () => {
+            const content = await TestServer.remoteAPI().getJSON('/api/test-groups');
```
This doesn't seem to not specify build request id at all??

rniwa commented on Websites/perf.webkit.org/server-tests/api-test-groups.js:
```diff
@@ -245,4 +246,55 @@ describe('/api/test-groups', function () {
+
+        it('should reject with "FailedToFetchTestGroups" if task id is invalid', async () => {
+            await MockData.addMockData(TestServer.database());
+            const content = await TestServer.remoteAPI().getJSON('/api/test-groups');
```
Ditto.

dewei-zhu replied:
Nice catch. I should be using assert.strictEqual below to actually run the test correctly.

dewei-zhu committed:
Extend test-groups API to be able to fetch test group information from a build request.
https://bugs.webkit.org/show_bug.cgi?id=270467
rdar://124024370

Reviewed by Ryosuke Niwa.

Add `/api/test-groups?buildRequest=<id>` API so that we can query test-group information from a build request.

* Websites/perf.webkit.org/public/api/test-groups.php: Added support for `/api/test-groups?buildRequest=<id>`.
* Websites/perf.webkit.org/server-tests/api-test-groups.js: Added unit tests for new API.

Canonical link: https://commits.webkit.org/275648@main
